src/algebraic_geometry/projective_spectrum/scheme.lean

```diff
@@ -120,36 +120,244 @@ def degree_zero_part {f : A} {m : ℕ} (f_deg : f ∈ 𝒜 m) : subring (away f)
+  simp only [mk_eq_mk', is_localization.eq] at eq1,
+  rcases eq1 with ⟨⟨_, ⟨M, rfl⟩⟩, eq1⟩,
+  erw [mul_one, mul_one] at eq1,
+  change f^_ * f^_ = _ * f^_ at eq1,
+  refine set.ne_empty_iff_nonempty.mpr ⟨f^N * f^M, eq1.symm ▸ mul_mem_right _ _
+    (sum_mem _ (λ i hi, mul_mem_left _ _ _)), ⟨N+M, by rw pow_add⟩⟩,
+  generalize_proofs h,
+  exact (classical.some_spec h).1,
 end
 
+/--The function between the basic open set `D(f)` in `Proj` to the corresponding basic open set in
+`Spec A⁰_f`. This is bundled into a continuous map in `Top_component.forward`.
+-/
+def to_fun (x : Proj.T| (pbo f)) : (Spec.T (A⁰_ f_deg)) :=
+⟨carrier f_deg x, carrier_ne_top f_deg x, λ x1 x2 hx12, begin
+  haveI : decidable_eq (away f) := classical.dec_eq _,
+  rcases ⟨x1, x2⟩ with ⟨⟨x1, hx1⟩, ⟨x2, hx2⟩⟩,
+  induction x1 using localization.induction_on with data_x1,
+  induction x2 using localization.induction_on with data_x2,
+  rcases ⟨data_x1, data_x2⟩ with ⟨⟨a1, _, ⟨n1, rfl⟩⟩, ⟨a2, _, ⟨n2, rfl⟩⟩⟩,
+  rcases mem_carrier.clear_denominator f_deg x hx12 with ⟨c, N, acd, eq1⟩,
+  simp only [degree_zero_part.coe_mul, algebra.smul_def] at eq1,
+  change localization.mk (f ^ N) 1 * (mk _ _ * mk _ _) = mk (∑ _, _) _ at eq1,
+  simp only [localization.mk_mul, one_mul] at eq1,
+  simp only [mk_eq_mk', is_localization.eq] at eq1,
+  rcases eq1 with ⟨⟨_, ⟨M, rfl⟩⟩, eq1⟩,
+  rw [submonoid.coe_one, mul_one] at eq1,
+  change _ * _ * f^_ = _ * (f^_ * f^_) * f^_ at eq1,
+
+  rcases x.1.is_prime.mem_or_mem (show a1 * a2 * f ^ N * f ^ M ∈ _, from _) with h1|rid2,
+  rcases x.1.is_prime.mem_or_mem h1 with h1|rid1,
+  rcases x.1.is_prime.mem_or_mem h1 with h1|h2,
+  { left,
+    rw mem_carrier_iff,
+    simp only [show (mk a1 ⟨f ^ n1, _⟩ : away f) = mk a1 1 * mk 1 ⟨f^n1, ⟨n1, rfl⟩⟩,
+      by rw [localization.mk_mul, mul_one, one_mul]],
+    exact ideal.mul_mem_right _ _ (ideal.subset_span ⟨_, h1, rfl⟩), },
+  { right,
+    rw mem_carrier_iff,
+    simp only [show (mk a2 ⟨f ^ n2, _⟩ : away f) = mk a2 1 * mk 1 ⟨f^n2, ⟨n2, rfl⟩⟩,
+      by rw [localization.mk_mul, mul_one, one_mul]],
+    exact ideal.mul_mem_right _ _ (ideal.subset_span ⟨_, h2, rfl⟩), },
+  { exact false.elim (x.2 (x.1.is_prime.mem_of_pow_mem N rid1)), },
+  { exact false.elim (x.2 (x.1.is_prime.mem_of_pow_mem M rid2)), },
+  { rw [mul_comm _ (f^N), eq1],
+    refine mul_mem_right _ _ (mul_mem_right _ _ (sum_mem _ (λ i hi, mul_mem_left _ _ _))),
+    generalize_proofs h,
+    exact (classical.some_spec h).1 },
+end⟩
+
+/-
+The preimage of basic open set `D(a/f^n)` in `Spec A⁰_f` under the forward map from `Proj A` to
+`Spec A⁰_f` is the basic open set `D(a) ∩ D(f)` in  `Proj A`. This lemma is used to prove that the
+forward map is continuous.
+-/
+lemma preimage_eq (a : A) (n : ℕ)
+  (a_mem_degree_zero : (mk a ⟨f ^ n, ⟨n, rfl⟩⟩ : away f) ∈ A⁰_ f_deg) :
+  to_fun 𝒜 f_deg ⁻¹'
+      (sbo (⟨mk a ⟨f ^ n, ⟨_, rfl⟩⟩, a_mem_degree_zero⟩ : A⁰_ f_deg)).1
+  = {x | x.1 ∈ (pbo f) ⊓ (pbo a)} :=
```

Comment:
There's a coercion missing from this statement, based on the `val_eq_coe`s below.

Reply:
I have eliminated all `subtype.val_eq_coe`

Reply:
The `x.1` part has to stay, but the `(sbo ...).1` can be eliminated